### PR TITLE
feat(pkg/scale): encoding and decoding of maps in scale

### PR DESCRIPTION
**Pull request by axaysagathiya**
## Changes

written the methods for encoding and decoding maps in scale.

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

#2796 

## Primary Reviewer

@timwu20

### CODE REVIEW

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:432]: nit rename `l` to `numberOfTuples`

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:434]: nit add error context so we can trace back what failed and which code path

```suggestion
		return fmt.Errorf("decoding length: %w", err)
```

**Review by kishansagathiya — COMMENTED** [review]
Fix this test `Test_unmarshal_optionality/testMap1 `

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:443]:
```suggestion
			return fmt.Errorf("decoding key %d of %d: %w", i+1, l, err)
```

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:450]:
```suggestion
			return fmt.Errorf("decoding value %d of %d: %w", i+1, l, err)
```

File: pkg/scale/encode.go
Comment (EclesioMeloJunior) [inline pkg/scale/encode.go:236]:
```suggestion
```

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:431]: nit rename `dstv` to `destination`

File: pkg/scale/decode.go
Comment (qdm12) [inline pkg/scale/decode.go:456]: nit to be explicit / the function is not that short

```suggestion
	return nil
```

File: pkg/scale/encode_test.go
Comment (EclesioMeloJunior) [inline pkg/scale/encode_test.go:1119]: you should pin the `tt` variable to the loop context otherwise it can cause inconsistency
```suggestion
	for _, tt := range mapTests {
	        tt := tt
		t.Run(tt.name, func(t *testing.T) {
```

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:232]:
```suggestion
		return fmt.Errorf("encoding length: %w", err)
```

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:237]:
```suggestion
```

File: pkg/scale/decode_test.go
Comment (EclesioMeloJunior) [inline pkg/scale/decode_test.go:156]: you should pin the tt variable to the loop context otherwise it can cause inconsistency
```suggestion
	for _, tt := range mapTests {
	        tt := tt
		t.Run(tt.name, func(t *testing.T) {
```

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:235]: nit rename `i` to `keyValue`, since `i` is usually for an index which isn't the case here :thinking:

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:241]:
```suggestion
			return fmt.Errorf("encoding map key: %w", err)
```

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:247]: Shouldn't we return an error in that case, to avoid silently discarding a map value when encoding?

@timwu20

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:251]:
```suggestion
			return fmt.Errorf("encoding map value: %w", err)
```

File: pkg/scale/encode.go
Comment (qdm12) [inline pkg/scale/encode.go:254]: nit

```suggestion
	return nil
```

**Review by qdm12 — COMMENTED** [review]
Nice, thanks :+1: 
Note one test is failing in the CI.

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]
Nice PR! I would like to suggest providing more test cases, for example a map from string to a struct, like:

```go
type exampleStruct struct {
  Acc string
  Value int
}

toEncode := map[string]exampleStruct{}
```

File: pkg/scale/decode.go
Comment (kishansagathiya) [reply pkg/scale/decode.go:431]: It is dstv everywhere in that file, there is no need to change this

File: pkg/scale/decode_test.go
Comment (qdm12) [reply pkg/scale/decode_test.go:156]: Only if `t.Parallel()` is used on the subtest :wink:

**Comment by jimjbrettj** [comment]
Seems rpc_03-chain_test is failing

File: pkg/scale/decode_test.go
Comment (jimjbrettj) [inline pkg/scale/decode_test.go:170]: Why are can we not treat maps like the other types here? Could we not define dst as the type of the input and unmarshal?

File: pkg/scale/encode.go
Comment (jimjbrettj) [reply pkg/scale/encode.go:253]: Def wait for Tims answer, but this happens at other places in this file as well (i.e. structs)

File: pkg/scale/encode_test.go
Comment (jimjbrettj) [inline pkg/scale/encode_test.go:916]: For both encoding and decoding, were any of the expected values derived from substrate? Or were the all derived from our own scale implementation?

**Review by jimjbrettj — COMMENTED** [review]
Great work, just have a few questions

File: pkg/scale/decode_test.go
Comment (axaysagathiya) [reply pkg/scale/decode_test.go:170]: if we use `dst := reflect.New(reflect.TypeOf(tt.in)).Interface()` for map,

- it panics in Unmarshal    
>  **panic: assignment to entry in nil map**

-   for `in` of type `map[int8][]byte`, it returns variable of type `*map[int8][]uint8` instead of `*map[int8][]byte`

so need to initialize the map and specify the type of `dst` same as type of `in`.

File: pkg/scale/decode_test.go
Comment (axaysagathiya) [reply pkg/scale/decode_test.go:136]: This is Rust code used to get expected encoded data.

File: pkg/scale/encode_test.go
Comment (axaysagathiya) [reply pkg/scale/encode_test.go:916]: yes all derived from substrate.

**Comment by axaysagathiya** [comment]
@qdm12  

Can you please rerun the test.
 Intermittent test `rpc_03-chain_test.go ` is  failing.

File: pkg/scale/decode_test.go
Comment (EclesioMeloJunior) [reply pkg/scale/decode_test.go:136]: @axaysagathiya I think it is not needed to put this comment block here, could you provide a link to the actual Rust code that does the encode?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: pkg/scale/decode_test.go
Comment (axaysagathiya) [reply pkg/scale/decode_test.go:136]: @EclesioMeloJunior this Rust code is written by me to get scale encoded data. Do you want me to provide a Gist link ?

Comment (EclesioMeloJunior) [reply pkg/scale/decode_test.go:136]: Not really, I mean your comment starts with: 

```
// // Rust code to encode a map of string to struct.
```

But actually, there is no code that encodes a map in your comment block, you're just adding two entries into a `BTreeMap`, that is why I would suggest removing this comment

File: pkg/scale/decode_test.go
Comment (jimjbrettj) [reply pkg/scale/decode_test.go:170]: cool cool that makes sense, thank you!

File: pkg/scale/decode_test.go
Comment (jimjbrettj) [reply pkg/scale/decode_test.go]: @kishansagathiya cool, I thought it was taken from somewhere else so I am fine to leave it as is. Want to avoid situations where we don't know where bytes came from.

Comment (EclesioMeloJunior) [reply pkg/scale/decode_test.go]: Oh, okay I did understand that you were trying to explain how rust encodes but you're explaining how you generated the encoded bytes that you're using in the test (my bad 😅)

Comment (kishansagathiya) [reply pkg/scale/decode_test.go]: > Oh, okay I did understand that you were trying to explain how rust encodes but you're explaining how you generated the encoded bytes that you're using in the test (my bad sweat_smile)

:laughing:

**Review by EclesioMeloJunior — APPROVED** [review]

File: pkg/scale/encode.go
Comment (timwu20) [reply pkg/scale/encode.go:253]: 'CanInterface` is essentially determining if this is a public attribute of a struct or public method.  I think it should be fine in this case.  This will almost always return true in the case we're trying to decode into a map value.  I wonder what cases this would return false though, maybe we can provide a test case.

@kishansagathiya please don't merge the PR with unresolved conversations.  My bad for not getting to this earlier.

File: pkg/scale/encode.go
Comment (axaysagathiya) [reply pkg/scale/encode.go:242]: yeah

File: pkg/scale/encode.go
Comment (kishansagathiya) [reply pkg/scale/encode.go:253]: I realised about this comment few moments after merging it. Wasn't deliberate.

File: pkg/scale/encode.go
Comment (timwu20) [reply pkg/scale/encode.go:242]: @axaysagathiya can you create another PR to only do this in the tests? This is not required when encoding maps in production, and it's not very performant.  I'd suggest adding a `sortKeys` parameter to the `encodeMap` function signature.

Comment (axaysagathiya) [reply pkg/scale/encode.go:242]: Oh, my bad! 
> This is not required when encoding maps

This is not just needed in test, this is needed in encoding as well. keys in go maps are unordered. so running encode function on the same go map could result into different encodings. Rust equivalent of go map is BTree map, where keys are sorted. Scale encode function in the Rust implementation does not produces different values on running it multiple times.

So, in order to make sure that our scale package and Rust's scale package encodes a map to the same value sorting is necessary.

Comment (timwu20) [reply pkg/scale/encode.go:242]: Just because the `BTreeMap` has deterministic sorting of keys, doesn't mean that the `parity-scale-codec` isn't able to decode maps that were encoded with keys in non-deterministic order.  In go it's understood that the order of the keys is non-deterministic.  In this case, we should be testing that we can encode a map and decode into a map, and assert that the maps are equal.  We could also ensure that decoding into a rust `BTreeMap` via the `parity-scale-codec` works as expected.  To make the tests more deterministic, you could provide a map with a single key or a map with two keys, and just assert that it's equal to one of the two encodings and remove the sorting logic altogether.

**Pull request merged**
